### ISSUE TITLE
A warning for the monomorphism restriction

### ISSUE THREAD

**Issue by noughtmare** [haskell/error-messages#56]
I have previously opened #40 which considers the case where an error is caused by the monomorphism restriction, however I think that the monomorphism restriction can be unexpected even if it doesn't cause other issues (e.g. [in this stack overflow question](https://stackoverflow.com/q/71950215/15207568)). So perhaps it would be a good idea to add a warning for functions that are affected by the monomorphism restriction. I guess the warning would be very similar to the defaulting warning, but only once per function and specifically mentioning the monomorphism restriction so that users can search for that keyword on the internet.

**Comment by goldfirere** [comment]
This already exists: `-Wmonomorphism-restriction`. Does that do what you want?

**Comment by noughtmare** [comment]
Ah, maybe it should be added to -Wall or is it there already too?

**Comment by goldfirere** [comment]
No, it's not in `-Wall`, and I don't think it should be added to `-Wall`: this will warn on every `x = 3` assignment in every program. Really, the monomorphism restriction is not an indication that anything is wrong -- it's just perhaps unexpected.

**Comment by noughtmare** [comment]
Ah, I was thinking that it would only warn if the monomorphism restriction actually had an observable effect (I believe that can only happen if it causes at least one type to be defaulted).

My second suggestion would be better then, namely changing the [`-Wtype-defaults`](https://downloads.haskell.org/~ghc/latest/docs/html/users_guide/using-warnings.html#ghc-flag--Wtype-defaults) warning to include a note about the monomorphism restriction if the type defaulting is caused by that.

**Comment by goldfirere** [comment]
Ah. That's an interesting idea. I really like that. So, to be concrete, right now, `x = 5` causes

```
Scratch.hs:32:5: warning: [-Wtype-defaults]
    * Defaulting the following constraint to type `Integer'
        Num a0 arising from the literal `5'
    * In the expression: 5
      In an equation for `x': x = 5
   |
32 | x = 5
   |     ^
```

But this idea suggests

```
Scratch.hs:32:5: warning: [-Wtype-defaults]
    * Defaulting the following constraint to type `Integer'
        Num a0 arising from the literal `5'
      This constraint could not be part of the context on 'x'
      because of the monomorphism restriction.
    * In the expression: 5
      In an equation for `x': x = 5
   |
32 | x = 5
   |     ^
```

What do we think?

**Comment by noughtmare** [comment]
I think that is an excellent error message. It captures all the required information and it is still brief.

**Comment by goldfirere** [comment]
https://gitlab.haskell.org/ghc/ghc/-/issues/21434
